### PR TITLE
removed juju-managed-units from app config

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

Remove `caas` config `juju-managed-units` entirely

## QA steps

```bash
juju deploy ./builds/mysql  --config juju-managed-units=true  # should raise error -> ERROR unknown option "juju-managed-units"
```

## Documentation changes

`Caas` is not live yet.

### CODE REVIEW

**Comment by ycliuhw** [comment]
[remove juju-managed-units from caas config](https://trello.com/c/ZcY5XRU2/225-remove-juju-managed-units-from-caas-config)

File: apiserver/facades/controller/caasunitprovisioner/provisioner.go [hunk 1]
```diff
@@ -217,18 +216,6 @@ func (a *Facade) UpdateApplicationsUnits(args params.UpdateApplicationUnitArgs)
 			result.Results[i].Error = common.ServerError(err)
 			continue
 		}
-
-		appConfig, err := app.ApplicationConfig()
```
Comment (wallyworld) [inline apiserver/facades/controller/caasunitprovisioner/provisioner.go]: The intent was to just remove the user facing aspect of this config, not the back end processing. There's logic besides this which would still need to change if the entire config were to be removed.

File: apiserver/facades/controller/caasunitprovisioner/provisioner_test.go [hunk 2]
```diff
@@ -189,15 +189,14 @@ func (s *CAASProvisionerSuite) TestApplicationConfig(c *gc.C) {
...
 	})
-	c.Assert(results.Results[0].Config, jc.DeepEquals, map[string]interface{}{"foo": "bar", "juju-managed-units": false})
+	c.Assert(results.Results[0].Config, jc.DeepEquals, map[string]interface{}{"foo": "bar"})
 }
 
 func strPtr(s string) *string {
 	return &s
 }
 
 func (s *CAASProvisionerSuite) TestUpdateApplicationsUnitsNoTags(c *gc.C) {
```
Comment (wallyworld) [inline apiserver/facades/controller/caasunitprovisioner/provisioner_test.go]: The tags vs no tags logic would all need to be fixed if we remove the juju-managed-units config from the backend entirely

File: worker/caasunitprovisioner/worker_test.go [hunk 3]
```diff
@@ -193,7 +193,6 @@ func (s *WorkerSuite) TestStartStop(c *gc.C) {
 }
 
 func (s *WorkerSuite) setupNewBrokerManagedUnitScenario(c *gc.C) worker.Worker {
```
Comment (wallyworld) [inline worker/caasunitprovisioner/worker_test.go]: And here, all these tests need fixing

File: worker/caasunitprovisioner/deployment_worker.go [hunk 4]
```diff
@@ -38,7 +37,7 @@ func newDeploymentWorker(
 ) (worker.Worker, error) {
 	w := &deploymentWorker{
 		application:        application,
-		jujuManagedUnits:   jujuManagedUnits,
```
Comment (wallyworld) [inline worker/caasunitprovisioner/deployment_worker.go]: This is deleting the setting of the attribute but there's all the business logic in the worker loop as well

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
!!build!!

File: worker/caasunitprovisioner/application_worker.go [hunk 5]
```diff
@@ -132,17 +128,7 @@ func (aw *applicationWorker) loop() error {
...
-			}
-			// Application has been removed, ensure all units are stopped
-			// before this worker is killed.
-			for unitId, w := range unitWorkers {
-				if err := aw.containerBroker.DeleteUnit(unitId); err != nil {
-					logger.Errorf("error deleting unit %v of removed application: %v", unitId, err)
-				}
-				worker.Stop(w)
-			}
+			continue
```
Comment (wallyworld) [inline worker/caasunitprovisioner/application_worker.go]: We can now just delete this channel, right?

Comment (wallyworld) [inline worker/caasunitprovisioner/application_worker.go]: I think the appRemoved channel can be deleted.

File: worker/caasunitprovisioner/deployment_worker.go [hunk 6]
```diff
@@ -138,20 +130,11 @@ func (w *deploymentWorker) loop() error {
 		if err != nil {
 			return errors.Annotate(err, "cannot parse pod spec")
 		}
-		// When Juju manages the units, we don't ask the
-		// substrate to create any itself.
-		if w.jujuManagedUnits {
-			numUnits = 0
-		}
 		err = w.broker.EnsureService(w.application, spec, numUnits, appConfig)
```
Comment (wallyworld) [inline worker/caasunitprovisioner/deployment_worker.go]: The EnsureService method has code that only creates a Deployment Controller if num units > 0.
That code can be simplified now. And there's also a check for num units >=0 which should now be just > 0

File: worker/caasunitprovisioner/application_worker.go (hunk 5 above)
Comment (ycliuhw) [reply worker/caasunitprovisioner/application_worker.go]: yes, agreed. remove it now.

File: caas/kubernetes/provider/k8s.go [hunk 7]
```diff
@@ -401,7 +401,7 @@ func (k *kubernetesClient) EnsureService(
 	logger.Debugf("creating/updating application %s", appName)
 
 	if numUnits < 0 {
-		return errors.Errorf("number of units must be >= 0")
+		return errors.Errorf("number of units must be > 0")
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: Above check should be <= 0 now

Comment (wallyworld) [reply caas/kubernetes/provider/k8s.go]: Also, the if numUnits > 0 below can disappear

Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: ah right

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**